extensions/amp-social-share/1.0/social-share.js

```diff
@@ -21,23 +21,36 @@ import {SocialShareIcon} from '../../../third_party/optimized-svg-icons/social-s
 import {addParamsToUrl, parseQueryString} from '../../../src/url';
 import {dict} from '../../../src/utils/object';
 import {getSocialConfig} from './amp-social-share-config';
-import {isObject} from '../../../src/types';
 import {openWindowDialog} from '../../../src/dom';
 import {useResourcesNotify} from '../../../src/preact/utils';
 
 const DEFAULT_WIDTH = 60;
 const DEFAULT_HEIGHT = 44;
+const DEFAULT_TARGET = '_blank';
 const NAME = 'SocialShare';
+
 /**
  * @param {!JsonObject} props
  * @return {PreactDef.Renderable}
  */
 export function SocialShare(props) {
   useResourcesNotify();
-  const {typeConfig, baseEndpoint, checkedWidth, checkedHeight} = checkProps(
-    props
+  const {
+    typeConfig,
+    baseEndpoint,
+    checkedWidth,
+    checkedHeight,
+    checkedTarget,
+    params,
+  } = checkProps(props);
+  let combinedParams = dict();
+  if (!props['ignoreParams']) {
+    combinedParams = {...typeConfig['defaultParams'], ...params};
+  }
+  const finalEndpoint = addParamsToUrl(
```

Comment:
So, this is very close, but I still don't quite like the idea of dumping any kind of args into the endpoint. I'd start even with no additional params whatsoever and instead, focus on the endpoint itself. IMHO, we should instead change our type config. Instead of providing a `typeConfig['shareEndpoint']` as a string and then interpolating it manually, why not just provide it as a function that would interpolate it directly based on props?

E.g. structurally it'd be like this:
```
{
  TYPEA: {
    endpoint: ({href, ...}) => `https://typeA.com/share?href=${encodeURIComponent(href)}&...`,
  },
}
```

This way we don't need to worry about cleaning up the props, or renaming query params. For instance, one type calls the shared URL a `url` and another `href` - this way we can just pick one prop name for URL and use it for every type.

The `endpoint` function will take care of everything (and it's also faster than anything else). And we'd simply call it like this:

```
function SocialShare(props) {
  ...
  const baseEndpoint = typeConfig[props.type].endpoint(props);
}
```

That will leave us only with two problems to solve for:
 1. Custom endpoint.
 2. Additional unforeseen params to pass to a URL.

/1/ Custom endpoint
Custom endpoint could be done exactly how you have it now. E.g. we emulate the `endpoint` function for it as `endpoint = (props) => addParamsToUrl(customEndpoint, paramsFromProps(props))`.

/2/ Additional unforeseen params to pass to a URL
Is that really a thing? If so, then we have two options: do as you're doing here, collect additional params, and add them to the URL. Or we could put them in the `additionalParams` prop as a map. Since they are indeed additional params, I like this option the best.

Put it all together, the code here could be exactly the same for all types:

```
const type = typeConfig[props.type];
let endpoint;
if (type) {
  endpoint = type.endpointFn(props);
} else {
  endpoint = addParamsToUrl(props.endpoint, {'href': props.href, ...});
}
if (props.additionalParams) {
  endpoint = addParamsToUrl(endpoint, props.additionalParams);
}
```

This also makes it pretty clear: standard things go to props and defined in the PropTypes, non-standard things go to the `additionalParams`, kind of like they do in the `data-param-` in the web component.

WDYT?

Reply:
Thanks for the thoughtful write up.  I do like the idea of having a custom function for each `type` and from this being able to have more standardized props.  So essentially we would have a few set type of props that would become params like:

- url/href
- text
- mini (seems to be special case for LinkedIn)
- recipient (seems to be special case for email)

I tried listing out how it might look in the above list.  So the idea would be that most of the types would provide a `url/href` and a `text` and then the few remaining ones would only need to be provided for that `type`?

Then we can use the customEndpoint to handle what has been passed in from the Amp side (after asynchronously resolving the bindings).

Yes, I think this approach works and gives us a cleaner setup for collecting props in preact.  My only concern is that the user may now be passing in a param that is different than the param name in the final endpoint.  i.e. if we name the `url` param `url` for all types, but in the actual endpoint it is `href` (like for facebook), the user might have a little confusion if they are used to 0.1 api.  I think if we document carefully, it should be fine.

**Additional Params**
I am also fine with passing in the additional params as a map, since they are extra params.  I do believe we should provide this functionality since it is available in the previous version of the component (0.1) and also now or in the future the social media sites might have additional functionality controlled by these extra params.

Reply:
I think the fact that some shares you `href` whereas our prop will be called `url` is totally ok. At the worst: `url` + `href` pretty much cover our range of variances here and we could always just do `props.url ?? props.href`. But i'd just start with one of these.

Reply:
sounds good, I ended up using 

`paramUrl` to cover both url and href.  Thanks for the feedback!